context/src/main/java/io/grpc/Context.java

```diff
@@ -109,37 +108,36 @@
    */
   public static final Context ROOT = new Context(null);
 
-  // One and only one of them is non-null
-  private static final Storage storage;
-  private static final LinkedBlockingQueue<ClassNotFoundException> storageOverrideNotFoundErrors =
-      new LinkedBlockingQueue<ClassNotFoundException>();
-  private static final Exception storageInitError;
-
-  static {
-    Storage newStorage = null;
-    Exception error = null;
-    try {
-      Class<?> clazz = Class.forName("io.grpc.override.ContextStorageOverride");
-      newStorage = (Storage) clazz.getConstructor().newInstance();
-    } catch (ClassNotFoundException e) {
-      // Avoid writing to logger because custom log handlers may try to use Context, which is
-      // problemantic (e.g., NullPointerException) because the Context class has not done loading
-      // at this point.
-      storageOverrideNotFoundErrors.add(e);
-      newStorage = new ThreadLocalContextStorage();
-    } catch (Exception e) {
-      error = e;
-    }
-    storage = newStorage;
-    storageInitError = error;
-  }
+  private static volatile Storage storage;
+  private static Exception storageInitError;
 
   // For testing
   static Storage storage() {
-    if (storage == null) {
-      throw new RuntimeException("Storage override had failed to initialize", storageInitError);
+    Storage tmp = storage;
+    if (tmp == null) {
+      synchronized (Context.class) {
+        tmp = storage;
+        if (tmp == null) {
+          if (storageInitError != null) {
+            throw new RuntimeException(
+                "Storage override had failed to initialize", storageInitError);
+          }
+          try {
+            Class<?> clazz = Class.forName("io.grpc.override.ContextStorageOverride");
+            tmp = (Storage) clazz.getConstructor().newInstance();
+            storage = tmp;
+          } catch (ClassNotFoundException e) {
+            tmp = new ThreadLocalContextStorage();
+            storage = tmp;
+            log.log(Level.FINE, "Storage override doesn't exist. Using default.", e);
+          } catch (Exception e) {
+            storageInitError = e;
```

Comment:
Without this field, next time it would fail but possibly would throw NPE from which user can not easily find the root cause if the original RuntimeExecption had happened quite long ago and only been logged.